cmd/otel-allocator/main.go

```diff
@@ -16,306 +16,205 @@ package main
+		promWatcherErr := promWatcher.Start(events, errors)
+		if promWatcherErr != nil {
+			setupLog.Error(promWatcherErr, "Failed to start prometheus watcher")
+			os.Exit(1)
+		}
+	}
+	srv := server.NewServer(log, allocator, targetDiscoverer, cliConf.ListenAddr)
+	interrupts := make(chan os.Signal, 1)
+	closer := make(chan bool, 1)
+	signal.Notify(interrupts, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+
+	runGroup.Add(
+		func() error {
+			fileWatcherErr := fileWatcher.Start(events, errors)
+			setupLog.Info("File watcher exited")
+			return fileWatcherErr
+		},
+		func(err error) {
+			setupLog.Info("Closing file watcher")
+			fileWatcherErr := fileWatcher.Close()
+			if fileWatcherErr != nil {
+				setupLog.Error(err, "file watcher failed to close")
+			}
+			setupLog.Info("Closing prometheus watcher")
+			promWatcherErr := promWatcher.Close()
+			if promWatcherErr != nil {
+				setupLog.Error(err, "prometheus watcher failed to close")
+			}
+		})
+	runGroup.Add(
+		func() error {
+			discoveryManagerErr := discoveryManager.Run()
+			setupLog.Info("Discovery manager exited")
+			return discoveryManagerErr
+		},
+		func(err error) {
+			setupLog.Info("Closing discovery manager")
+			discoveryCancel()
+		})
+	runGroup.Add(
+		func() error {
+			// Initial loading of the config file's scrape config
+			err = targetDiscoverer.ApplyConfig(allocatorWatcher.EventSourceConfigMap, cfg.Config)
+			if err != nil {
+				setupLog.Error(err, "Unable to apply initial configuration")
+				return err
+			}
+			err := targetDiscoverer.Watch(allocator.SetTargets)
+			setupLog.Info("Target discoverer exited")
+			return err
+		},
+		func(err error) {
+			setupLog.Info("Closing target discoverer")
+			targetDiscoverer.Close()
+		})
+	runGroup.Add(
+		func() error {
+			err := collectorWatcher.Watch(ctx, cfg.LabelSelector, allocator.SetCollectors)
+			setupLog.Info("Collector watcher exited")
+			return err
+		},
+		func(err error) {
+			setupLog.Info("Closing collector watcher")
+			collectorWatcher.Close()
+		})
+	runGroup.Add(
+		func() error {
+			err := srv.Start()
+			setupLog.Info("Server failed to start")
+			return err
+		},
+		func(err error) {
+			setupLog.Info("Closing server")
 			if err := srv.Shutdown(ctx); err != nil {
 				setupLog.Error(err, "Error on server shutdown")
-				os.Exit(1)
 			}
-			os.Exit(0)
-		case event := <-watcher.Events:
-			eventsMetric.WithLabelValues(event.Source.String()).Inc()
-			switch event.Source {
-			case allocatorWatcher.EventSourceConfigMap:
-				setupLog.Info("ConfigMap updated!")
-				// Restart the server to pickup the new config.
-				if err := srv.Shutdown(ctx); err != nil {
-					setupLog.Error(err, "Cannot shutdown the server")
-				}
-				srv = newServer(log, allocator, discoveryManager, k8sclient, cliConf.ListenAddr)
-				go func() {
-					if err := srv.Start(); !errors.Is(err, http.ErrServerClosed) {
-						setupLog.Error(err, "Can't restart the server")
+		})
+	runGroup.Add(
+		func() error {
```

Comment:
The run function here seems to have multiple responsibilities, but the best summary I can think of is "channel watcher".  I'm not as familiar with the run package so I'm not sure if it's normal to have a "catch all" function or if it would be better to split this up into two functions: one for interrupt logic and one for handler logic.